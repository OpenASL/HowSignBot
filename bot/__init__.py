__version__ = "21.12.1"
