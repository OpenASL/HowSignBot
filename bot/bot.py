import asyncio
import itertools
import logging

import disnake
from disnake import ApplicationCommandInteraction
from disnake.ext import commands

from . import settings

logger = logging.getLogger(__name__)

COMMAND_PREFIX = settings.COMMAND_PREFIX


intents = disnake.Intents.default()
intents.typing = False
intents.presences = False
intents.bans = False
intents.integrations = False
intents.webhooks = False
intents.invites = False
intents.members = True
intents.messages = True
bot = commands.Bot(
    command_prefix=COMMAND_PREFIX,
    case_insensitive=True,
    owner_id=settings.OWNER_ID,
    intents=intents,
    sync_commands_debug=settings.DEBUG,
    reload=settings.DEBUG,
    test_guilds=settings.TEST_GUILDS or None,
)

PRESENCE_COMMANDS = (
    "/sign",
    "/handshape",
    "/catchphrase | /codenames",
    "/schedule new",
    "/watch2gether",
    "/sentence | /idiom",
    "/feedback",
)
CHANGE_PRESENCE_EVERY = 120  # seconds


async def cycle_presence():
    for command in itertools.cycle(PRESENCE_COMMANDS):
        activity = disnake.Activity(
            name=command,
            type=disnake.ActivityType.watching,
        )
        logger.debug(f"changing presence to show command: {command}")
        await bot.change_presence(activity=activity)
        await asyncio.sleep(CHANGE_PRESENCE_EVERY)


@bot.event
async def on_ready():
    bot.loop.create_task(cycle_presence())


@bot.event
async def on_command_error(ctx, error):
    if isinstance(
        error,
        (commands.errors.CheckFailure, commands.errors.BadArgument),
    ):
        await ctx.send(error)
    else:
        logger.error(f"unhandled exception from command: {ctx.command}", exc_info=error)


@bot.event
async def on_slash_command_error(inter: ApplicationCommandInteraction, error: Exception):
    if isinstance(
        error,
        (commands.errors.CheckFailure, commands.errors.BadArgument),
    ):
        await inter.send(error)
    else:
        logger.error(
            f"unhandled exception from command: {inter.application_command.name!r}",
            exc_info=error,
        )


async def set_default_presence():
    activity = disnake.Activity(
        name="/sign",
        type=disnake.ActivityType.watching,
    )
    await bot.change_presence(activity=activity)
