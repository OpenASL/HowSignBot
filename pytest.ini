[pytest]
filterwarnings =
    ignore:.*@coroutine.*:DeprecationWarning
