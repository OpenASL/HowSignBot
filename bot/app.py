import asyncio
import logging

import aiohttp_cors
from aiohttp import web
from ariadne import graphql
from ariadne.constants import PLAYGROUND_HTML

from . import settings
from .bot import bot
from .database import store
from .graphql.schema import schema
from .utils.extensions import walk_extensions

logger = logging.getLogger(__name__)

# Assign app to bot so that extensions can add routes
bot.app = app = web.Application()  # type: ignore

app.cors = cors = aiohttp_cors.setup(
    app,
    defaults={
        "*": aiohttp_cors.ResourceOptions(
            allow_credentials=False,
            expose_headers="*",
            allow_headers="*",
        )
    },
)


async def ping(_):
    return web.Response(body="", status=200)


def graphql_playground(_):
    return web.Response(text=PLAYGROUND_HTML, status=200, content_type="text/html")


async def graphql_server(request):
    data = await request.json()

    success, result = await graphql(
        schema, data, context_value=request, debug=settings.DEBUG
    )

    status = 200 if success else 400
    return web.json_response(result, status=status)


app.add_routes([web.get("/ping", ping)])
resource = bot.app.router.add_resource("/graphql")  # type: ignore
cors.add(resource.add_route("POST", graphql_server))


async def start_bot():
    try:
        await bot.start(settings.DISCORD_TOKEN)
    finally:
        await bot.close()


async def on_startup(app):
    for ext in walk_extensions():
        bot.load_extension(ext)
    await store.connect()
    app["bot_task"] = asyncio.create_task(start_bot())
    app["bot"] = bot


async def on_shutdown(app):
    app["bot_task"].cancel()
    await app["bot_task"]
    await store.disconnect()


app.on_startup.append(on_startup)
app.on_shutdown.append(on_shutdown)
