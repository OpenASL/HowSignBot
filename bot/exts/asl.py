import logging
from urllib.parse import quote_plus

import discord
from discord.ext.commands import Cog, Bot, Context, command, errors

import handshapes
from bot import settings
from bot.utils import get_spoiler_text, did_you_mean

logger = logging.getLogger(__name__)

COMMAND_PREFIX = settings.COMMAND_PREFIX

SIGN_TEMPLATE = """[🤲 **Handspeak** - Search results]({handspeak})
[🧬 **Lifeprint** - Search results]({lifeprint})
[🤝 **SigningSavvy** - Sign for {word_uppercased}]({signingsavvy})
[🌐 **Spread The Sign** - {word_uppercased}]({spread_the_sign})
[📹 **YouGlish** - ASL videos with {word_uppercased}]({youglish})
Share: {howsign}
"""

SIGN_SPOILER_TEMPLATE = """[🤲 **Handspeak** - Search results]({handspeak})
[🧬 **Lifeprint** - Search results]({lifeprint})
[🤝 **SigningSavvy** - Sign for ||{word_uppercased}||]({signingsavvy})
[🌐 **Spread The Sign** - ||{word_uppercased}||]({spread_the_sign})
[📹 **YouGlish** - ASL videos with ||{word_uppercased}||]({youglish})
Share: ||{howsign}||
"""

SIGN_HELP = """Look up a word or phrase

If the word or phrase is sent in spoiler text, i.e. enclosed in `||`, the word will also be blacked out in the reply.
To search multiple words/phrases, separate the values with a comma.

Examples:
{COMMAND_PREFIX}sign tiger
{COMMAND_PREFIX}sign ||tiger||
{COMMAND_PREFIX}sign what's up
{COMMAND_PREFIX}sign church, chocolate, computer
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def word_display(word: str, *, template: str = SIGN_TEMPLATE, max_length: int = 100):
    if len(word) > max_length:
        raise errors.BadArgument("⚠️ Input too long. Try a shorter query.")
    quoted_word = quote_plus(word).lower()
    return template.format(
        word_uppercased=word.upper(),
        howsign=f"https://howsign.app/?s={quoted_word}",
        lifeprint=f"https://www.google.com/search?&q=site%3Alifeprint.com+{quoted_word}",
        handspeak=f"https://www.google.com/search?&q=site%3Ahandspeak.com+{quoted_word}",
        signingsavvy=f"https://www.signingsavvy.com/search/{quoted_word}",
        spread_the_sign=f"https://www.spreadthesign.com/en.us/search/?q={quoted_word}",
        youglish=f"https://youglish.com/pronounce/{quoted_word}/signlanguage/asl",
    )


def sign_impl(word: str):
    logger.info(f"sending links for: '{word}'")
    spoiler = get_spoiler_text(word)
    word = spoiler if spoiler else word
    has_multiple = "," in word
    template = SIGN_SPOILER_TEMPLATE if spoiler else SIGN_TEMPLATE
    if has_multiple:
        words = word.split(",")
        embed = discord.Embed()
        for word in words:
            word = word.strip()
            title = f"||{word.upper()}||" if spoiler else word.upper()
            embed.add_field(name=title, value=word_display(word, template=template))
    else:
        title = f"||{word.upper()}||" if spoiler else word.upper()
        embed = discord.Embed(
            title=title,
            description=word_display(word, template=template),
        )

    return {"embed": embed}


HANDSHAPE_HELP = """Show a random or specific handshape

Examples:
{COMMAND_PREFIX}handshape
{COMMAND_PREFIX}handshape claw5

Enter {COMMAND_PREFIX}handshapes to show a list of handshapes.
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def handshape_impl(name: str):
    logger.info(f"handshape: '{name}'")
    try:
        if name == "random":
            handshape = handshapes.get_random_handshape()
        else:
            handshape = handshapes.get_handshape(name)
    except handshapes.HandshapeNotFoundError:
        logger.info(f"handshape '{name}' not found")
        suggestion = did_you_mean(name, tuple(handshapes.HANDSHAPES.keys()))
        if suggestion:
            return {
                "content": f'"{name}" not found. Did you mean "{suggestion}"? Enter `{COMMAND_PREFIX}handshapes` to see a list of handshapes.'
            }
        else:
            return {
                "content": f'"{name}" not found. Enter `{COMMAND_PREFIX}handshapes` to see a list of handshapes.'
            }

    filename = f"{handshape.name}.png"
    file_ = discord.File(handshape.path, filename=filename)
    embed = discord.Embed(title=handshape.name)
    embed.set_image(url=f"attachment://{filename}")
    return {
        "file": file_,
        "embed": embed,
    }


HANDSHAPES_TEMPLATE = """{handshapes}

Enter `{COMMAND_PREFIX}handshape` to display a random handshape or `{COMMAND_PREFIX}handshape [name]` to display a specific handshape.
"""

HANDSHAPES_HELP = """List handshapes

Enter {COMMAND_PREFIX}handshape to display a random handshape or {COMMAND_PREFIX}handshape [name] to display a specific handshape.
""".format(
    COMMAND_PREFIX=COMMAND_PREFIX
)


def handshapes_impl():
    return {
        "content": HANDSHAPES_TEMPLATE.format(
            handshapes=", ".join(handshapes.HANDSHAPES.keys()),
            COMMAND_PREFIX=COMMAND_PREFIX,
        )
    }


class ASL(Cog):
    def __init__(self, bot: Bot):
        self.bot = bot

    @command(name="sign", aliases=("howsign", COMMAND_PREFIX), help=SIGN_HELP)
    async def sign_command(self, ctx: Context, *, word: str):
        await ctx.send(**sign_impl(word))

    @sign_command.error
    async def sign_error(self, ctx: Context, error: Exception):
        # Ignore "??"
        if isinstance(error, errors.MissingRequiredArgument):
            if ctx.invoked_with == COMMAND_PREFIX:
                logger.info(
                    f"no argument passed to {COMMAND_PREFIX}{ctx.invoked_with}. ignoring..."
                )
            else:
                await ctx.send(
                    f"⚠️ Enter a word or phrase to search for after `{COMMAND_PREFIX}{ctx.invoked_with}`."
                )

    @command(name="handshape", aliases=("shape",), help=HANDSHAPE_HELP)
    async def handshape_command(self, ctx: Context, name="random"):
        await ctx.send(**handshape_impl(name))

    @command(name="handshapes", aliases=("shapes",), help=HANDSHAPES_HELP)
    async def handshapes_command(self, ctx: Context):
        logger.info("sending handshapes list")
        await ctx.send(**handshapes_impl())


def setup(bot: Bot) -> None:
    bot.add_cog(ASL(bot))
