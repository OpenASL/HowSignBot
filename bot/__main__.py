import logging

from aiohttp import web

from . import __version__, settings
from .app import app

log_format = "%(asctime)s - %(name)s %(levelname)s: %(message)s"
logging.getLogger("discord").setLevel(logging.WARNING)
logging.basicConfig(
    format=log_format,
    level=settings.LOG_LEVEL,
)
logger = logging.getLogger(__name__)

logger.info(f"starting bot version {__version__}")
web.run_app(app, port=settings.PORT)
