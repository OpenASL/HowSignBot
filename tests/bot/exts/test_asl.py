import random

import pytest
from disnake.ext import commands
from syrupy.filters import props

from bot.exts import asl

random.seed(1)


@pytest.mark.parametrize(
    "word",
    (
        "tiger",
        "||tiger||",
        "what's up",
        "||what's up||",
        "need, ask",
        "||need, ask||",
    ),
)
def test_sign(snapshot, word):
    result = asl.sign_impl(word)
    assert result == snapshot


def test_sign_long_input():
    with pytest.raises(commands.errors.BadArgument, match="too long"):
        asl.sign_impl("a" * 101)


@pytest.mark.parametrize("name", ("random", "open8", "open9"))
def test_handshape(snapshot, name):
    result = asl.handshape_impl(name)
    assert result == snapshot(
        exclude=props(
            "fp",
        )
    )


def test_handshapes(snapshot):
    result = asl.handshapes_impl(prefix="!")
    assert result == snapshot
