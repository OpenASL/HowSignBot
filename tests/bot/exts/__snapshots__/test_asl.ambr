# name: test_handshape[open8]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=Embed.Empty,
      colour=Embed.Empty,
      description=Embed.Empty,
      fields=<class 'list'> [
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(url='attachment://Open8.png'),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title='Open8',
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
    'file': <class 'File'> {
      description=None,
      filename='Open8.png',
      spoiler=False,
    },
  }
---
# name: test_handshape[open9]
  <class 'dict'> {
    'content': '"open9" not found. Did you mean "OpenX"? Enter `!handshapes` to see a list of handshapes.',
  }
---
# name: test_handshape[random]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=Embed.Empty,
      colour=Embed.Empty,
      description=Embed.Empty,
      fields=<class 'list'> [
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(url='attachment://8.png'),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title='8',
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
    'file': <class 'File'> {
      description=None,
      filename='8.png',
      spoiler=False,
    },
  }
---
# name: test_handshapes
  <class 'dict'> {
    'content': '1, 3, Bent3, 4, 5, Claw5, 6, 7, 8, Open8, 9, Flat9, A, OpenA, B, BentB, FlatB, OpenB, C, FlatC, SmallC, D, E, G, H, I, K, L, M, OpenM, N, OpenN, O, FlatO, SmallO, R, S, T, V, BentV, X, OpenX, Y, ILY, Corna',
  }
---
# name: test_sign[need, ask]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=Embed.Empty,
      colour=Embed.Empty,
      description=Embed.Empty,
      fields=<class 'list'> [
        EmbedProxy(inline=True, name='NEED', value='[🤲 **Handspeak** - Search results](https://www.google.com/search?&q=site%3Ahandspeak.com+need)\n[🧬 **Lifeprint** - Search results](https://www.google.com/search?&q=site%3Alifeprint.com+need)\n[🤝 **SigningSavvy** - Sign for NEED](https://www.signingsavvy.com/search/need)\n[🌐 **Spread The Sign** - NEED](https://www.spreadthesign.com/en.us/search/?q=need)\n[📜 **ASL Signbank** - NEED](https://aslsignbank.haskins.yale.edu/signs/search/?keyword=need)\n[📹 **YouGlish** - ASL videos with NEED](https://youglish.com/pronounce/need/signlanguage/us)\nShare: https://howsign.app/?s=need\n'),
        EmbedProxy(inline=True, name='ASK', value='[🤲 **Handspeak** - Search results](https://www.google.com/search?&q=site%3Ahandspeak.com+ask)\n[🧬 **Lifeprint** - Search results](https://www.google.com/search?&q=site%3Alifeprint.com+ask)\n[🤝 **SigningSavvy** - Sign for ASK](https://www.signingsavvy.com/search/ask)\n[🌐 **Spread The Sign** - ASK](https://www.spreadthesign.com/en.us/search/?q=ask)\n[📜 **ASL Signbank** - ASK](https://aslsignbank.haskins.yale.edu/signs/search/?keyword=ask)\n[📹 **YouGlish** - ASL videos with ASK](https://youglish.com/pronounce/ask/signlanguage/us)\nShare: https://howsign.app/?s=ask\n'),
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title=Embed.Empty,
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
  }
---
# name: test_sign[tiger]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=Embed.Empty,
      colour=Embed.Empty,
      description='
        [🤲 **Handspeak** - Search results](https://www.google.com/search?&q=site%3Ahandspeak.com+tiger)
        [🧬 **Lifeprint** - Search results](https://www.google.com/search?&q=site%3Alifeprint.com+tiger)
        [🤝 **SigningSavvy** - Sign for TIGER](https://www.signingsavvy.com/search/tiger)
        [🌐 **Spread The Sign** - TIGER](https://www.spreadthesign.com/en.us/search/?q=tiger)
        [📜 **ASL Signbank** - TIGER](https://aslsignbank.haskins.yale.edu/signs/search/?keyword=tiger)
        [📹 **YouGlish** - ASL videos with TIGER](https://youglish.com/pronounce/tiger/signlanguage/us)
        Share: https://howsign.app/?s=tiger
  
      ',
      fields=<class 'list'> [
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title='TIGER',
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
  }
---
# name: test_sign[what's up]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=Embed.Empty,
      colour=Embed.Empty,
      description='
        [🤲 **Handspeak** - Search results](https://www.google.com/search?&q=site%3Ahandspeak.com+what%27s+up)
        [🧬 **Lifeprint** - Search results](https://www.google.com/search?&q=site%3Alifeprint.com+what%27s+up)
        [🤝 **SigningSavvy** - Sign for WHAT'S UP](https://www.signingsavvy.com/search/what%27s+up)
        [🌐 **Spread The Sign** - WHAT'S UP](https://www.spreadthesign.com/en.us/search/?q=what%27s+up)
        [📜 **ASL Signbank** - WHAT'S UP](https://aslsignbank.haskins.yale.edu/signs/search/?keyword=what%27s+up)
        [📹 **YouGlish** - ASL videos with WHAT'S UP](https://youglish.com/pronounce/what%27s+up/signlanguage/us)
        Share: https://howsign.app/?s=what%27s+up
  
      ',
      fields=<class 'list'> [
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title="WHAT'S UP",
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
  }
---
# name: test_sign[||need, ask||]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=Embed.Empty,
      colour=Embed.Empty,
      description=Embed.Empty,
      fields=<class 'list'> [
        EmbedProxy(inline=True, name='||NEED||', value='[🤲 **Handspeak** - Search results](https://www.google.com/search?&q=site%3Ahandspeak.com+need)\n[🧬 **Lifeprint** - Search results](https://www.google.com/search?&q=site%3Alifeprint.com+need)\n[🤝 **SigningSavvy** - Sign for ||NEED||](https://www.signingsavvy.com/search/need)\n[🌐 **Spread The Sign** - ||NEED||](https://www.spreadthesign.com/en.us/search/?q=need)\n[📜 **ASL Signbank** - ||NEED||](https://aslsignbank.haskins.yale.edu/signs/search/?keyword=need)\n[📹 **YouGlish** - ASL videos with ||NEED||](https://youglish.com/pronounce/need/signlanguage/us)\nShare: ||https://howsign.app/?s=need||\n'),
        EmbedProxy(inline=True, name='||ASK||', value='[🤲 **Handspeak** - Search results](https://www.google.com/search?&q=site%3Ahandspeak.com+ask)\n[🧬 **Lifeprint** - Search results](https://www.google.com/search?&q=site%3Alifeprint.com+ask)\n[🤝 **SigningSavvy** - Sign for ||ASK||](https://www.signingsavvy.com/search/ask)\n[🌐 **Spread The Sign** - ||ASK||](https://www.spreadthesign.com/en.us/search/?q=ask)\n[📜 **ASL Signbank** - ||ASK||](https://aslsignbank.haskins.yale.edu/signs/search/?keyword=ask)\n[📹 **YouGlish** - ASL videos with ||ASK||](https://youglish.com/pronounce/ask/signlanguage/us)\nShare: ||https://howsign.app/?s=ask||\n'),
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title=Embed.Empty,
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
  }
---
# name: test_sign[||tiger||]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=Embed.Empty,
      colour=Embed.Empty,
      description='
        [🤲 **Handspeak** - Search results](https://www.google.com/search?&q=site%3Ahandspeak.com+tiger)
        [🧬 **Lifeprint** - Search results](https://www.google.com/search?&q=site%3Alifeprint.com+tiger)
        [🤝 **SigningSavvy** - Sign for ||TIGER||](https://www.signingsavvy.com/search/tiger)
        [🌐 **Spread The Sign** - ||TIGER||](https://www.spreadthesign.com/en.us/search/?q=tiger)
        [📜 **ASL Signbank** - ||TIGER||](https://aslsignbank.haskins.yale.edu/signs/search/?keyword=tiger)
        [📹 **YouGlish** - ASL videos with ||TIGER||](https://youglish.com/pronounce/tiger/signlanguage/us)
        Share: ||https://howsign.app/?s=tiger||
  
      ',
      fields=<class 'list'> [
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title='||TIGER||',
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
  }
---
# name: test_sign[||what's up||]
  <class 'dict'> {
    'embed': <class 'Embed'> {
      Empty=Embed.Empty,
      author=EmbedProxy(),
      color=Embed.Empty,
      colour=Embed.Empty,
      description='
        [🤲 **Handspeak** - Search results](https://www.google.com/search?&q=site%3Ahandspeak.com+what%27s+up)
        [🧬 **Lifeprint** - Search results](https://www.google.com/search?&q=site%3Alifeprint.com+what%27s+up)
        [🤝 **SigningSavvy** - Sign for ||WHAT'S UP||](https://www.signingsavvy.com/search/what%27s+up)
        [🌐 **Spread The Sign** - ||WHAT'S UP||](https://www.spreadthesign.com/en.us/search/?q=what%27s+up)
        [📜 **ASL Signbank** - ||WHAT'S UP||](https://aslsignbank.haskins.yale.edu/signs/search/?keyword=what%27s+up)
        [📹 **YouGlish** - ASL videos with ||WHAT'S UP||](https://youglish.com/pronounce/what%27s+up/signlanguage/us)
        Share: ||https://howsign.app/?s=what%27s+up||
  
      ',
      fields=<class 'list'> [
      ],
      footer=EmbedProxy(),
      image=EmbedProxy(),
      provider=EmbedProxy(),
      thumbnail=EmbedProxy(),
      timestamp=Embed.Empty,
      title="||WHAT'S UP||",
      type='rich',
      url=Embed.Empty,
      video=EmbedProxy(),
    },
  }
---
